### PR TITLE
Define some simple string types to build upon

### PR DESCRIPTION
**Pull request by jrha**
* Non-empty string
* String without any white-space
* String without leading or trailing white-space

### CODE REVIEW

File: pan/strings.pan [hunk 1]
```diff
@@ -0,0 +1,22 @@
+declaration template pan/strings;
+
+
+@documentation{
+Convenience type specifying a string that must contain at least one character (anything other than an empty string).
+Intended for use in building more complex types.
+}
+type string_non_empty = string with match(SELF, '^.+$');
```
Comment (stdweird) [inline pan/strings.pan]: `'.'` is enough, you can also test the lenght of the string; that probably even faster

File: pan/strings.pan [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+
+@documentation{
+Convenience type specifying a string that may not contain any whitespace.
+Intended for use in building more complex types.
+}
+type string_non_whitespace = string with match(SELF, '^\S+$');
```
Comment (stdweird) [inline pan/strings.pan]: use `!match(SELF, '\s')`?

**Comment by jrha** [comment]
Feedback on names welcome, before they become spread all over the place.

File: pan/strings.pan [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+Intended for use in building more complex types.
+}
+type string_non_empty = string(1..);
+
+
+@documentation{
+Convenience type specifying a string that may not contain any whitespace.
+Intended for use in building more complex types.
+}
+type string_non_whitespace = string with !match(SELF, '\s');
```
Comment (stdweird) [inline pan/strings.pan]: this allows empty string, not sure if you want that. if it is allowed, maybe clarify in documnetation

File: pan/strings.pan [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+
+@documentation{
+Convenience type specifying a string that may contain whitespace, but may not start or end with whitespace.
```
Comment (stdweird) [inline pan/strings.pan]: regexp doesn't allow empty string, documentation shoudl reflect that

File: pan/strings.pan (hunk 3 above)
Comment (jrha) [reply pan/strings.pan]: Wasn't supposed to be allowed before your suggested change, will fix.

File: pan/strings.pan [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+}
+type string_non_whitespace = string_non_empty with !match(SELF, '\s');
+
+
+@documentation{
+Convenience type specifying a string that must contain at least one character, may contain whitespace, but may not
+start or end with whitespace.
+Intended for use in building more complex types.
+}
+type string_trimmed = string_non_empty with match(SELF, '^(?=\S).+(?<=\S)$');
```
Comment (stdweird) [inline pan/strings.pan]: this regexp looks very complex, for something that is probably `^\S(.*\S)?$'

Comment (jrha) [reply pan/strings.pan]: Probably. I can't spend any more time on this now though.

**Comment by jrha** [comment]
@stdweird updated, I hope the regexs now meet with your approval :suspect:

File: pan/strings.pan [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+@documentation{
+A string that must contain at least one character (anything other than an empty string).
+}
+type string_non_empty = string(1..);
+
+
+@documentation{
+A string that must contain at least one character and may not contain any whitespace.
+}
+type string_non_whitespace = string_non_empty with !match(SELF, '\s');
```
Comment (jouvin) [inline pan/strings.pan]: `with match(SELF, '\S')`?

Comment (jrha) [reply pan/strings.pan]: No, that would mean "at least one non-whitespace character"

**Comment by jrha** [comment]
`\S` → Matches anything other than a space, tab or newline.

File: pan/strings.pan (hunk 6 above)
Comment (jouvin) [reply pan/strings.pan]: you are right! I wanted to say: `with match(SELF, '^\S+$')` . But probably yours is more efficient, agreed.

File: pan/strings.pan [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+A string that must contain at least one character, may contain whitespace, but may not start or end with whitespace.
+}
+type string_trimmed = string_non_empty with match(SELF, '^\S(.*\S)?$');
+
+
+@documentation{
+A string that must contain at least one character and no uppercase characters
+}
+type string_lowercase = string_non_empty with ! match(SELF, '[[:upper:]]');
+
```
Comment (jouvin) [inline pan/strings.pan]: Base on `string_trimmed` rather than `string_non_empty`?

File: pan/strings.pan [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+@documentation{
+A string that must contain at least one character and no lowercase characters
+}
+type string_uppercase = string_non_empty with ! match(SELF, '[[:lower:]]');
```
Comment (jouvin) [inline pan/strings.pan]: Same remark as for `string_lowercase`

File: pan/strings.pan (hunk 7 above)
Comment (jrha) [reply pan/strings.pan]: Could do, any objections?

**Comment by jrha** [comment]
If you want to be more restrictive, choose from:
`\w` →Matches any letter, digit or underscore. Equivalent to [a-zA-Z0-9_].
`[[:alnum:]]` →POSIX class of the above, potentially allows for Unicode matching.

**Comment by jouvin** [comment]
My point is about using something more restrictive than `^\S+$` for `non_empty` and the others derived from it is that I expect these types to be used in circumstances (e.g. file names) where you don't wnat non printable characters at least. May be `[[:alnum:]]` but I think it is missing `-` at least...

**Comment by jrha** [comment]
`[[:alnum:]-._]`?

**Comment by jouvin** [comment]
About https://github.com/quattor/template-library-core/pull/180#issuecomment-394689678: I don't have an example on the top of my mind but I can imagine it to be useful. Basic idea: no whitespace, trimmed, First character uppercase, others lowercase. But we can add it when we have a use case...

File: pan/strings.pan [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+@documentation{
+A string that must contain at least one character and may not contain any whitespace.
+}
+type string_non_whitespace = string_non_empty with !match(SELF, '\s');
+
+
+@documentation{
+A string that must contain at least one character, may contain whitespace, but may not start or end with whitespace.
+}
+type string_trimmed = string_non_empty with match(SELF, '^\S(.*\S)?$');
```
Comment (ned21) [inline pan/strings.pan]: I am not sure about this name. It's important that something that appears in a file far away from where it's defined is obvious to the reader. Does trimming something always mean whitespace? A quick google shows excel and jquery functions that do so maybe it is obvious.

Comment (jrha) [reply pan/strings.pan]: I'm keen to make sure these names make sense, so very happy to take suggestions.

AFAIK most languages call the function that does this either `trim` or `strip` and `trim` seems to be slightly more common.

Comment (jrha) [reply pan/strings.pan]: FWIW a quick survey gave me:

Language | Function Name
--|--
.NET | `Trim()`
Arduino | `trim()`
Excel | `trim()`
Go | `Trim()`
Java | `trim()`
JS | `trim()`
PHP | `trim()`
Python  | `strip()`
R | `trimws()`
Ruby | `strip()`

**Comment by jrha** [comment]
@ned21 @stdweird are you both happy with the names?

**Comment by ned21** [comment]
@jrha yes.

**Pull request merged**